rust/main/submitter/src/payload_dispatcher/building_stage.rs

```diff
@@ -23,20 +27,113 @@ struct BuildingStage {
 }
 
 impl BuildingStage {
-    pub async fn run(&self) -> Result<()> {
+    pub async fn run(&self) {
         loop {
-            let payload = self.queue.lock().await.pop_front();
-            if let Some(payload) = payload {
-                let txs = self.state.adapter.build_transactions(vec![payload]).await?;
-                for tx in txs {
-                    self.inclusion_stage_sender.send(tx).await?;
+            // event-driven by the Building queue
+            let payload = match self.queue.lock().await.pop_front() {
+                Some(payload) => payload,
+                None => {
+                    // wait for the next payload to arrive
+                    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+                    continue;
+                }
+            };
+
+            let payloads = vec![payload];
+            let txs = match self
+                .state
+                .adapter
+                .build_transactions(payloads.clone())
+                .await
+            {
+                Err(err) => {
+                    error!(
+                        ?err,
+                        ?payloads,
+                        "Error building transactions. Dropping payloads"
+                    );
+                    let details_for_payloads: Vec<_> =
+                        payloads.into_iter().map(|p| p.details().clone()).collect();
+                    self.drop_payloads(&details_for_payloads, DropReason::UnhandledError)
+                        .await;
+                    continue;
+                }
+                Ok(txs) => txs,
+            };
+
+            for tx in txs {
+                if let Err(err) = self.simulate_tx(&tx).await {
+                    error!(?err, "Error simulating transaction. Dropping transaction");
+                    // TODO: distinguish between network error and failed simulation
+                    self.drop_tx(&tx, DropReason::FailedSimulation).await;
+                    continue;
+                };
+                if let Err(err) = self.send_tx_to_inclusion_stage(tx.clone()).await {
+                    error!(?err, "Error storing transaction in the database");
+                    self.drop_tx(&tx, DropReason::UnhandledError).await;
+                    continue;
+                } else {
+                    self.store_tx(&tx).await;
                 }
-            } else {
-                // wait for the next payload to arrive
-                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
             }
         }
     }
+
+    async fn drop_payloads(&self, details: &[PayloadDetails], reason: DropReason) {
+        for d in details {
+            if let Err(err) = self
+                .state
+                .payload_db
+                .set_payload_status(d.id(), PayloadStatus::Dropped(reason.clone()))
+                .await
+            {
+                error!(?err, "Error updating payload status to `dropped`");
```

Comment:
I would add payload details into this log.

Reply:
done